### PR TITLE
chore: cancel context on signals

### PR DESCRIPTION
**Pull request by alecthomas**

### CODE REVIEW

File: internal/server/signals.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	"syscall"
+
+	"github.com/TBD54566975/ftl/internal/log"
+)
+
+// RunWithSignalHandler installs a signal handler that cancels the context on SIGINT and SIGTERM, sends a SIGTERM to the
+// process group, and exits.
+func RunWithSignalHandler(ctx context.Context, run func(ctx context.Context) error) error {
+	logger := log.FromContext(ctx)
+	ctx, stop := signal.NotifyContext(ctx, syscall.SIGINT, syscall.SIGTERM)
```
Comment (stuartwdouglas) [inline internal/server/signals.go:17]: Doesn't the work around context cancellation happen asynchronously? Given that we immediately send SIGTERM won't it be non-deterministic as to which stuff actually sees the context cancellation?

Comment (alecthomas) [reply internal/server/signals.go:17]: "run" won't return until it and all its child goroutines are complete, at which point

Comment (stuartwdouglas) [reply internal/server/signals.go:17]: What makes 'run' not return until child goroutines complete?

Comment (alecthomas) [reply internal/server/signals.go:17]: The design of the function... using errgroups to track sub-tasks, etc...

Comment (alecthomas) [reply internal/server/signals.go:17]: Correct, this is just the introduction of the basic cancellation.

Actually I may not be able to merge this. Turns out the only way the BuildEngine will terminate currently is by being literally killed, so yep.

**Pull request closed without merging**